docs/design/data-protection-test.md

```diff
@@ -0,0 +1,219 @@
+Components involved and their responsibilities:
+
+- **DataProtectionTest (DPT) CRD:**
+    - **Spec:**
+        - **backupLocation:** Contains Velero-based backup storage configuration.
+        - **uploadSpeedTestConfig:** Test parameters for the upload (file size, test timeout).
+        - **CSIVolumeSnapshotTestConfig:** Test parameters for the CSI VolumeSnapshot test (snapshot class, source PVC name, and importantly the PVC namespace, plus snapshot timeout).
+    - **Status:**
+        - **UploadTestStatus:** Groups upload speed (in Mbps), success flag, and error messages.
+        - **SnapshotTestStatus:** Groups the snapshot test results, reporting status and the duration taken for the snapshot to be ready.
+        - **S3Vendor:** Reports the detected S3 vendor string from vendor determination.
+- **DataProtectionTest Controller:**
+    - Monitors DataProtectionTest CRs.
+    - Extracts configuration from the backup location spec.
+    - Determines the S3 vendor via an HTTP HEAD call.
+    - Initializes the appropriate cloud provider using the CloudProvider interface.
+    - Executes the upload test and, if enabled, the CSI VolumeSnapshot test.
+    - Updates the CRD status with grouped results.
+- **CloudProvider Interface:**
+    - Defines an `UploadTest(ctx, config, bucket, fileSizeMB) (int64, error)` function.
+    - AWS-specific implementation (S3Provider) is provided using the AWS SDK.
+- **Vendor Determination Logic:**
+    - A helper function performs an HTTP HEAD call to the **s3Url** and inspects headers (especially the `Server` header) to determine the vendor (e.g., "AWS", "MinIO", etc.).
+
+## Detailed Design
+
+#### Proposed DataProtectionTest CRD:
+
+DataProtectionTest (DPT) CRD would look like:
+
+```yaml
+apiVersion: oadp.openshift.io/v1alpha1
+kind: DataProtectionTest
+metadata:
+  name: my-data-protection-test
+spec:
+  backupLocation:
+    provider: aws                # Cloud provider type (aws, azure, gcp)
+    default: true
+    objectStorage:
+      bucket: sample-bucket
+      prefix: velero
+    config:
+      region: us-east-1
+      profile: "default"
+      insecureSkipTLSVerify: "true"
+      s3Url: "https://s3.amazonaws.com" # indicates s3 compatibility
+    credential:
+      name: cloud-credentials     # Secret for cloud credentials
+      key: cloud
+  uploadSpeedTestConfig:
+    fileSize: "100MB"                # Size of file to upload for testing
+    testTimeout: "60s"               # Maximum duration for upload test
+  CSIVolumeSnapshotTestConfigs:
+    - snapshotClassName: "csi-snapclass"
+      volumeSnapshotSource:
+        persistentVolumeClaimName: "db-pvc-1"
+        persistentVolumeClaimNamespace: "app1"
+      timeout: "60s"
+    - snapshotClassName: "csi-snapclass"
+      volumeSnapshotSource:
+        persistentVolumeClaimName: "db-pvc-2"
+        persistentVolumeClaimNamespace: "app1"
+      timeout: "60s"
+status:
+  lastTested: "2024-10-08T10:00:00Z"
+  uploadTest:
+    speedMbps: 55.3
```

Comment:
Should we add totalMB and duration here too?

Reply:
size is already specified by the user, in the spec. So just duration ?

Reply:
Added duration.

Reply:
Might be easier to see all 3 together -- and I don't know if it's possible for the transferred bytes to be different than specified, but it's not essential. Just adding duration is fine too.